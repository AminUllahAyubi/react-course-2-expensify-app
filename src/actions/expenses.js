
import uuid from "uuid"
// ADD_EXPENSE
export const addExpense=(
    {
        id=uuid(),
        description='',
        note='',
        amount=0,
        createAt=0
    
    }={})=>{
    return{
        type:"ADD_EXPENSE",
        expense:{
            id:uuid(),
            description,
            note,
            amount,
            createAt
        }
    }
}

// REMVOE_EXPENSE
export const removeExpense=({id})=>{
    return{
        type:'REMOVE_EXPENSE',
        id
    }
}
// EDITE_EXPENSE
export const editeExpense=(id,update)=>{
    return{
        type:'EDITE_EXPENSE',
        id,
        update
    }
}

