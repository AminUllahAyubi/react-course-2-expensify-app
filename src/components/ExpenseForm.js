
import React from "react";
import { connect, ReactReduxContext } from "react-redux";
import { SingleDatePicker } from 'react-dates';
// import 'react-dates/lib/css/_datepicker.css';
import moment from 'moment';
const now = moment();
console.log(now.format('MMM Do YYYY'))
class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            // createAt:moment(),
            createAt: props.expense ? props.expense.createAt : new Date(),
            error: '',
            // calenderFocused:false
        }
    }
    onDescriptionChange = (e) => {
        let description = e.target.value;
        this.setState(() => {
            return {
                description
            }
        })
    }
    onNoteChange = (e) => {
        let note = e.target.value;
        e.persist()
        this.setState(() => {
            return {
                note
            }
        })
    }
    onAmountChange = (e) => {
        let amount = e.target.value;
        e.persist();
        const reg = /^\d{1,}(\.\d{0,2})?$/;
        if (!amount || amount.match(reg)) {
            this.setState(() => {
                return {
                    amount
                }
            })
        }
    }
    onDateChange = (e) => {
        let createAt = e.target.value;
        e.persist();
        this.setState(() => {
            return {
                createAt
            }
        })
    }
    // onDateChange=(createAt)=>{
    //     this.setState(()=>({createAt}))
    // }
    // onFocusChange=({focused})=>({calenderFocused:focused})

    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => {
                return { error: 'Please fill the description and amount' }
            })
        }
        else {
            this.setState(() => {
                return { error: '' }
            })
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createAt: this.state.createAt.valueOf(),
                note: this.state.note
            })
        }
    }
    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={(e) => {
                            this.onDescriptionChange(e)
                        }}
                    >
                    </input>
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={(e) => {
                            this.onAmountChange(e)
                        }}
                    >
                    </input>
                    {/* css and sass is not completly configure when it done use the bellow component */}
                    {/* <SingleDatePicker
                        date={this.state.createAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calenderFocused}
                        onFocusChange={this.onFocusChange}
                    /> */}
                    <input type='date'
                        onClick={(e) => { e.target.type = 'date' }}
                        value={this.state.createAt}
                        onChange={(e) => { this.onDateChange(e) }}
                    >
                    </input>
                    <textarea
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={(e) => {
                            this.onNoteChange(e);
                        }}
                    >
                    </textarea>
                    <button >Add Expense</button>
                </form>
            </div>
        )
    }
}
export default ExpenseForm