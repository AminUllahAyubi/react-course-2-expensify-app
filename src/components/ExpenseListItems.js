import React, { createElement } from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
export default ({ id, description, note, amount, createAt}) => {
    return (
        <div>
            <h3>
                <Link to={`/edit#${id}`}>{description}</Link></h3>
            <p>{amount} - {createAt}</p>
        </div>
    )
}
// const mapStateToState = (state) => {
//     return {
//         expenses: state.expenses
//     }
// }
// export default connect(mapStateToState)(ExpenseListItems);
// export default ExpenseListItems;

