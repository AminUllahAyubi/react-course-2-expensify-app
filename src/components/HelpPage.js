import React from "react";
  const HelpPage = () => {
    return (
      <div>
        <p>This is Help Page</p>
      </div>
    );
  };
  export default HelpPage;