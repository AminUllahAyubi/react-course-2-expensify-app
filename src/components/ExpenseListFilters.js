import React from "react";
import { connect } from "react-redux";
import { setTextFilter, sortByAmount, sortByDate } from '../actions/filters'
const ExpenseListFilters = (props) => (
    <div>
        <input type="text" value={props.filters.text} onChange={(e) => {
            props.dispatch(setTextFilter({ text: e.target.value }))
       }}></input>
        <select
            value={props.filters.sortBy}
            
            onChange={(e) => {
                if (e.target.value === 'date') {
                    props.dispatch(sortByDate());
                }
                else if (e.target.value === 'amount') {
                    props.dispatch(sortByAmount())
                }
            }}>
            <option value="date">date</option>
            <option value="amount">Amount</option>
        </select>
    </div>
);
const mapPropsToState = (state) => {
    return {
        filters: state.filters
    }
}

const ConnectExpenseListFilters = connect(mapPropsToState)(ExpenseListFilters)

export default ConnectExpenseListFilters;