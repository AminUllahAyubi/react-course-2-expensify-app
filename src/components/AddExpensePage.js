import React from "react";
// import expenses from "../reducers/expenses";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { addExpense  } from "../actions/expenses";
const addExpensePage = (props) => {
    return (
      <div>
        <h2>Add Expense</h2>
        <ExpenseForm
          onSubmit={(expense)=>{
            props.dispatch(addExpense(expense))
            props.history.push('/')
          }}
        />
      </div>
    );
  };
export default connect()(addExpensePage);
