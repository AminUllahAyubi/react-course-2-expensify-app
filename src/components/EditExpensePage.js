import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { editeExpense } from "../actions/expenses";
import { removeExpense } from "../actions/expenses";
const EditExpensePage = (props) => {
  return (
    <div>
      <ExpenseForm
        expense={props.expense}
        onSubmit={(expense) => {
          props.dispatch(editeExpense(props.location.hash.slice(1), expense));
          props.history.push('/');
        }}
      />
      <button onClick={() => {
        props.dispatch(removeExpense({ id: props.location.hash.slice(1) }))
        props.history.push('/');
      }}>Remove</button>
    </div>
  );
};
const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => {
      return expense.id == props.location.hash.slice(1)
    })
  };
}
export default connect(mapStateToProps)(EditExpensePage);
