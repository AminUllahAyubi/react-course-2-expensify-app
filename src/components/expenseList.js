import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItems from './ExpenseListItems';
import selectExpense from '../selectors/expenses'
export const ExpenseList = (props) => (
    <p>
        {
            props.expenses.length=== 0 ? (
                <p>No Expenses Exsists</p>
            ) : (
                props.expenses.map((expense) => {
                    return (
                        <ExpenseListItems
                            {...expense}
                            key={expense.id}
                        />
                    )
                })
            )
        }
    </p>
);


const mapStateToProps = (state) => {
    return {
        expenses: selectExpense(state.expenses, state.filters)
    }
}
const ConnectExpenseList = connect(mapStateToProps)(ExpenseList);
export default ConnectExpenseList