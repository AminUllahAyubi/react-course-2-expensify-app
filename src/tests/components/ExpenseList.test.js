import React from 'react';
import {shallow} from 'enzyme';
import { ExpenseList } from '../../components/expenseList';
import expenses from '../fixtures/expenses';
import ExpenseListItems from '../../components/ExpenseListItems';
test('should test the expense list component with expenses list data',()=>{
    const wrapper=shallow(<ExpenseList expenses={expenses}/>);
    expect(wrapper).toMatchSnapshot();
})
test('should test the expense list component without expenses list data',()=>{
    const wrapper=shallow(<ExpenseList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
})
