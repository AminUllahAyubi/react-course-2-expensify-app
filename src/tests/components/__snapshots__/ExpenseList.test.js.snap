// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test the expense list component with expenses list data 1`] = `
<p>
  <Component
    amount={90}
    createAt={0}
    description="Water"
    id="1"
    key="1"
    note=""
  />
  <Component
    amount={90}
    createAt={234230}
    description="Gas bill"
    id="2"
    key="2"
    note=""
  />
  <Component
    amount={990900}
    createAt={2340}
    description="Foods bill"
    id="3"
    key="3"
    note=""
  />
</p>
`;

exports[`should test the expense list component without expenses list data 1`] = `
<p>
  <p>
    No Expenses Exsists
  </p>
</p>
`;
